tmt/steps/provision/connect.py

```diff
@@ -31,12 +31,103 @@ class ConnectGuestData(tmt.steps.provision.GuestSshData):
         help='Username to use for all guest operations.'
         )
 
+    soft_reboot: Optional[ShellScript] = field(
+        default=None,
+        option='--soft-reboot',
+        metavar='COMMAND',
+        help="""
+             If specified, the command, executed on the runner, would be used
+             for soft reboot of the guest.
+             """,
+        normalize=tmt.utils.normalize_shell_script,
+        serialize=lambda value: str(value) if isinstance(value, ShellScript) else None,
+        unserialize=lambda serialized: None if serialized is None else ShellScript(serialized)
+        )
+    hard_reboot: Optional[ShellScript] = field(
+        default=None,
+        option='--hard-reboot',
+        help="""
+             If specified, the command, executed on the runner, would be used
+             for hard reboot of the guest.
+             """,
+        metavar='COMMAND',
+        normalize=tmt.utils.normalize_shell_script,
+        serialize=lambda value: str(value) if isinstance(value, ShellScript) else None,
+        unserialize=lambda serialized: None if serialized is None else ShellScript(serialized)
+        )
+
 
 @dataclasses.dataclass
 class ProvisionConnectData(ConnectGuestData, tmt.steps.provision.ProvisionStepData):
     pass
 
 
+class GuestConnect(tmt.steps.provision.GuestSsh):
+    _data_class = ConnectGuestData
+
+    soft_reboot: Optional[ShellScript]
+    hard_reboot: Optional[ShellScript]
+
+    def reboot(
+            self,
+            hard: bool = False,
+            command: Optional[Union[Command, ShellScript]] = None,
+            timeout: Optional[int] = None,
+            tick: float = tmt.utils.DEFAULT_WAIT_TICK,
+            tick_increase: float = tmt.utils.DEFAULT_WAIT_TICK_INCREASE) -> bool:
+        """
+        Reboot the guest, and wait for the guest to recover.
+
+        :param hard: if set, force the reboot. This may result in a loss of
+            data. The default of ``False`` will attempt a graceful reboot.
+        :param command: a command to run on the guest to trigger the reboot.
+            If not set, plugin would try to use
+            :py:attr:`ConnectGuestData.soft_reboot` or
+            :py:attr:`ConnectGuestData.hard_reboot` (``--soft-reboot`` and
+            ``--hard-reboot``, respectively), if specified. Unlike ``command``,
+            these would be executed on the runner, **not** on the guest.
+        :param timeout: amount of time in which the guest must become available
+            again.
+        :param tick: how many seconds to wait between two consecutive attempts
+            of contacting the guest.
+        :param tick_increase: a multiplier applied to ``tick`` after every
+            attempt.
+        :returns: ``True`` if the reboot succeeded, ``False`` otherwise.
+        """
+
+        if not command:
+            if hard and self.hard_reboot is not None:
+                self.debug(f"Reboot using the command '{self.hard_reboot}'.")
```

Comment:
```suggestion
                self.debug(f"Reboot using the hard reboot command '{self.hard_reboot}'.")
```

Perhaps? To make it clear in the log whether soft or hard option was used. Similarly below.

Reply:
Fixed in be4a36bb